### ISSUE TITLE
Smarter handling of markdown headings in entry

### ISSUE THREAD

**Issue by jjatria** [maebert/jrnl#335]
If I write a long entry formatted in Markdown, I would like to write something like

~~~~
2015-03-18 13:42 A long entry
# First part

This is the first part, and it can go on like this for a while.

# Second part

This is the second part, where the premise of the first part gets all
crooked and interesting!

# Third part

Now everything gets wrapped up nicely and I profit. Somehow.
~~~~

I expect that so far nobody thinks this is a particularly badly formatted entry. But this causes all sorts of problems, because if I do `jrnl --export markdown` I'll get the following output:

~~~~
2015
====

March
-----

### 2015-03-19 20:50, 2015-03-18 13:42 A long entry 

# First part

This is the first part, and it can go on like this for a while.

# Second part

This is the second part, where the premise of the first part gets all
crooked and interesting!

# Third part

Now everything gets wrapped up nicely and I profit. Somehow.
~~~~

And that is most certainly not right. I understand why this is the case, but this is surely not how it _should_ be. I imagine the correct output should be something like this:

~~~~
2015
====

March
-----

### 2015-03-19 20:50, 2015-03-18 13:42 A long entry 

#### First part

This is the first part, and it can go on like this for a while.

#### Second part

This is the second part, where the premise of the first part gets all
crooked and interesting!

#### Third part

Now everything gets wrapped up nicely and I profit. Somehow.
~~~~

I'm not sure how difficult it would be to generate this last output, but I think that is what `jrnl --export markdown` _should_ generate. That is certainly what I expected it to generate.

Any thoughts on this?

**Comment by maebert** [comment]
Ah, so basically, you'd want to "increase the indentation" (ie. header depth) of all headers in your journal entry by 3?

Definitely for this. Putting this out for grabs, maybe @flight16, @notbalanced @pcarranza, @MinchinWeb, @aniketpant or any of the other usual suspects want to take this on?

**Comment by aniketpant** [comment]
@maebert I think what @jjatria is saying is quite correct. The header hierarchy is wrong in this kind of jrnl entry. But there's a small caveat here, jrnl isn't made for writing markdown. For jrnl any entry is essentially plain text with only a simple support for tags.

What do you think?

**Comment by jjatria** [comment]
@aniketpant is right in that jrnl should not assume markdown in its entries. But maybe this is the kind of cases in which an option is useful: if you somehow (in the config? as an option?) tell jrnl that the entry _is_ in markdown, then maybe each entry could be parsed _before_ the export, and then manipulated to fit in with the rest of the exported text.

**Comment by flight16** [comment]
I agree with @jjatria that this could be a journal-level option.  A MarkdownJournal Journal type associated with .md files would also be slick as a separate enhancement.

**Comment by MinchinWeb** [comment]
I've been thinking about doing something similar for a while now.

I think it is fine to assume that the entry text is markdown when you are exporting your journal to markdown.

Implementation questions:

1. does this need a switch in settings to be able to turn it on and off?
2. does this need a setting to control how many levels the headings are increased?
3. if exporting to one markdown files, how many levels to increase the headings?
3. if exporting to individual markdown files, how many levels to increase the headings?
4. what should we do if increasing the heading levels puts them past `h6`?

I propose the following answers:

1. No.
2. No.
3. Three.
3. One.
4. Ignore the error.

**Comment by flight16** [comment]
I agree with @MinchinWeb's on implementation answers, with the small addition that it would be friendly to warn to STDERR in case 5 when increasing the headings would put us past level h6.

**Comment by aniketpant** [comment]
After reading @MinchinWeb's comment, the implementation looks trivial to me. So, let's take a final call on this. What say @maebert?

**Comment by maebert** [comment]
@MinchinWeb Excellent run-down, agree with all five points. @aniketpant and @flight16 Go for it!

**Comment by pcarranza** [comment]
@maebert @MinchinWeb  +1

I was also thinking about handling the "bold" case when printing to the terminal if you want to write markdown in your jrnl. But I think that it is beyond scope for what is actually being requested for here.

**Comment by jjatria** [comment]
This is great. I'm looking forward to see how it turns out!

**Comment by aniketpant** [comment]
@maebert Will get started on this today and submit a PR.

**Comment by maebert** [comment]
@aniketpant Awesome, thank you! The `2.0-rc1` branch should be mostly error-free now!

**Comment by aniketpant** [comment]
I might take a few days to get started on this. If someone wants to pick it up, then please feel free to do so.

**Comment by MinchinWeb** [comment]
Here it is!

Implemented as per my suggestions, plus @flight16 's suggestion for a warning if we go past `h6`. I couldn't seem to find a library that would change Markdown heading levels, so I used regex instead. Works on both ATX and Setext style headings, while leaving `hr`'s in place.

I also fixed a bug where slugs for multi-file export always started with a 'b' when using Python 3.

I also fixed a couple of code documentation things while I was at it.

Tested here on Windows, and passes the `behave` tests on Travis-CI.
